In [3]:
import requests
from bs4 import BeautifulSoup as bs, Tag
import re
import pandas as pd
import time

In [4]:
url = 'https://en.wikipedia.org/wiki/List_of_largest_private_non-governmental_companies_by_revenue'
page = requests.get(url)
response = requests.get(url)
html = response.text
status = response.status_code
if status == 200:
    page = response.text
    soup = bs(page)
else:
    print(f"Oops! Received status code {status}")

In [5]:
table = soup.find_all('table')[0]

In [6]:
headers = table.find_all('th')

In [7]:
headers = [title.text.strip() for title in headers]
headers

['No.', 'Company', 'Revenue (in billions of USD)', 'Headquarters', 'Founded']

In [8]:
df = pd.DataFrame(columns = headers)
df

Empty DataFrame
Columns: [No., Company, Revenue (in billions of USD), Headquarters, Founded]
Index: []

In [9]:
data_row = table.find_all('tr')


In [10]:
def clean_citations(text):
    return re.sub(r'\[\d+\]', '', text).strip()

cleaned_data = []

for row in data_row[1:]:
    if isinstance(row, str):
        parsed = bs(row, 'html.parser')
    elif isinstance(row, Tag):
        parsed = row
    else:
        continue

    row_data = parsed.find_all('td')
    table_rows = [clean_citations(data.text) for data in row_data]
    cleaned_data.append(table_rows)

print(cleaned_data)

[['1', 'Vitol', '400 (2023)', 'Switzerland', '1966'], ['2', 'Trafigura Group', '244.3 (2023)', 'Singapore', '1993'], ['3', 'Schwarz Group', '180.6 (2023)', 'Germany', '1930'], ['4', 'Cargill', '160 (2023/24)', 'United States', '1865'], ['5', 'Koch Industries', '125 (2021)', 'United States', '1940'], ['6', 'Aldi', '120.96 (2023)', 'Germany', '1946'], ['7', 'E.Leclerc', '102.98 (2023)', 'France', '1948'], ['8', 'Suning Holdings Group', '102.9 (2019/20)', 'China', '1990'], ['9', 'REWE Group', '99.69 (2023)', 'Germany', '1927'], ['10', 'Huawei', '99 (2023)', 'China', '1987'], ['11', 'Robert Bosch', '98.92 (2023)', 'Germany', '1886'], ['12', 'Amer International Group', '88.9 (2019)', 'China', '1995'], ['13', 'HNA Group', '88.5 (2018)', 'China', '2000'], ['14', 'ByteDance', '85.2 (2022) *', 'China', '2012'], ['15', 'State Farm', '82.2 (2021)', 'United States', '1922'], ['16', 'Hengli Group', '80.6 (2019)', 'China', '1994'], ['17', 'Pacific Construction Group', '77.07 (2021)', 'China', '1986'

In [11]:
for row in cleaned_data:
    df.loc[len(df)] = row

In [12]:
df

No.                         Company Revenue (in billions of USD)  \
0      1                           Vitol                   400 (2023)   
1      2                 Trafigura Group                 244.3 (2023)   
2      3                   Schwarz Group                 180.6 (2023)   
3      4                         Cargill                160 (2023/24)   
4      5                 Koch Industries                   125 (2021)   
..   ...                             ...                          ...   
118  119                             G4S                  10.3 (2024)   
119  120  Jones Financial (Edward Jones)                 10.17 (2020)   
120  121                          Hy-Vee                   10.1 (2019   
121  122                         Calpine                  10.1 (2019)   
122  123                    Pacific Life                 10.06 (2020)   

       Headquarters Founded  
0       Switzerland    1966  
1         Singapore    1993  
2           Germany    1930  
3     United States    1865  
4     United States    1940  
..              ...     ...  
118  United Kingdom    2004  
119   United States    1922  
120   United States    1930  
121   United States    1984  
122   United States    1868  

[123 rows x 5 columns]

In [13]:
df.to_csv(r'C:\Users\pmush\OneDrive\Desktop\Data Science Prog\webscrapping\largeprivate.csv', index = False)

In [14]:
df = pd.read_csv('C:/Users/pmush/OneDrive/Desktop/Data Science Prog/webscrapping/largeprivate.csv')

In [30]:
counts = df['Headquarters'].value_counts()

print(counts)

Headquarters
United States     62
China             18
Germany           10
France            10
United Kingdom    10
Switzerland        3
Japan              3
Netherlands        3
Singapore          1
Sweden             1
South Korea        1
Spain              1
Name: count, dtype: int64


## Question 5 ##
#### What percentage are American companies in the largest private non-governmental companies? ####

In [32]:
percentages = df['Headquarters'].value_counts(normalize=True) * 100

print(percentages)

Headquarters
United States     50.406504
China             14.634146
Germany            8.130081
France             8.130081
United Kingdom     8.130081
Switzerland        2.439024
Japan              2.439024
Netherlands        2.439024
Singapore          0.813008
Sweden             0.813008
South Korea        0.813008
Spain              0.813008
Name: proportion, dtype: float64


### Answer ###
* American companies make up 50% of the 123 Largest private non-governmental companies listed by revenue
* the page was last edited on 8 April 2025

### References ###
count and percentage code: ChatGPT, response to “how to get count and percentage for one column,” April 12, 2025, ChatGPT, https://chatgpt.com/.

clean citations code: ChatGPT, response to “how to remove citations from a table,” April 11, 2025, ChatGPT, https://chatgpt.com/.